nameko/containers.py

```diff
@@ -127,9 +139,14 @@ def __init__(self, service_cls, worker_ctx_cls, config):
         self.max_workers = (
             config.get(MAX_WORKERS_CONFIG_KEY) or DEFAULT_MAX_WORKERS)
 
-        self.dependencies = DependencySet()
-        for dep in prepare_dependencies(self):
-            self.dependencies.add(dep)
+        # the relationship between extensions and the container is symbiotic;
+        # containers are responsible for starting and stopping their
+        # extensions, and extensions are able to interact with the container
+        # via its :attr:`ServiceContainer.interface`.
+        self.dependencies = ExtensionSet()
```

Comment:
the `ExtensionSet` class is a bit strange to me. in `discover_extensions` we combine providers and entrypoints, only to put them in a set with methods that loop over them to split them up again...?